src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs

```diff
@@ -161,22 +161,19 @@ private async Task<string> DownloadFromUrlAsync(IHostEnvironment env, IChannel c
                     return (await t).Message;
                 }
 
-                return CheckValidDownload(ch, filePath, url, ref deleteNeeded);
+                return CheckValidDownload(filePath, url, ref deleteNeeded);
             }
         }
 
-        private static string CheckValidDownload(IChannel ch, string filePath, string url, ref bool deleteNeeded)
+        private static string CheckValidDownload(string filePath, string url, ref bool deleteNeeded)
         {
             // If the relative url does not exist, aka.ms redirects to www.microsoft.com. Make sure this did not happen.
-            // If the file is big then it is definitely not the redirect.
             var info = new FileInfo(filePath);
-            if (info.Length > 4096)
-                return null;
```

Comment:
I'm worried without a size check, a dataset consisting of HTML would not work. For instance when classifying news articles scraped from the news websites. The existing size check ensured that larger datasets were not blocked. 

We can subclass `WebClient` to track the final URI (after redirects): ([idea src](http://www.nullskull.com/a/1484/make-the-webclient-class-follow-redirects-and-get-target-url.aspx))
```c#

public class WebClientResponseUri : WebClient
{
	public Uri ResponseUri { get; private set; }

	protected override WebResponse GetWebResponse(WebRequest request)
	{
		WebResponse response = null;
		
		try
		{
			response = base.GetWebResponse(request);
			ResponseUri = response.ResponseUri;
		}
		catch
		{
		}
		
		return response;
	}
}
```

Then use to check the URI:
```c#
private Exception DownloadResource(IHostEnvironment env, IChannel ch, WebClientResponseUri /* Note new class name */ webClient, Uri uri, string path, string fileName, CancellationToken ct)
{
    ...
    using (var ws = fh.CreateWriteStream())
    {
        var headers = webClient.ResponseHeaders.GetValues("Content-Length");
        var requestUri = webClient.ResponseUri; /* -- start new part -- */
   
        if (requestUri.Host == "www.microsoft.com" && requestUri.ToString().Length < 60)
            return ch.Except($"Invalid aka.ms url: {uri}");  /* -- end new part --*/
    
        if (Utils.Size(headers) == 0 || !long.TryParse(headers[0], out var size))
            size = 10000000;
        ...
```

Have to modify `DownloadFromUrlWithRetryAsync()`:
```C#
        private async Task<string> DownloadFromUrlWithRetryAsync(IHostEnvironment env, IChannel ch, string url, string fileName,
            int timeout, string filePath, int retryTimes = 5)
		{
			var downloadResult = "";

			for (int i = 0; i < retryTimes; ++i)
			{
				var thisDownloadResult = await DownloadFromUrlAsync(env, ch, url, fileName, timeout, filePath);

				if (string.IsNullOrEmpty(thisDownloadResult))
					return thisDownloadResult;
				else
				{
					downloadResult += thisDownloadResult + @"\n";

					// Do not retry if the URL does not exist
					if (thisDownloadResult.Contains("does not exist"))
						return downloadResult;
				}

				await Task.Delay(10 * 1000);
			}

			return downloadResult;
		}
```

Then can remove `CheckValidDownload()`.

Tested on .NET Fiddle -- https://dotnetfiddle.net/5zs4c2 (can modify `url` in a correct value to see non-redirect case)

Perhaps could benefit from a unit test which tests a bad aka.ms link?

Reply:
These APIs cannot be used to download arbitrary datasets, they are only used to (internally) download resources from specific locations (namely, urls under [https://aka.ms/mlnet-resources/](https://aka.ms/mlnet-resources/)). So I don't think it is necessary to go to the trouble of making this relatively big change.

I do agree with the last sentence in your comment though, I remember that we had a bunch of tests for the `ResourceManagerUtils` class in TLC, not sure where they are.

---
In reply to: [392572981](https://github.com/dotnet/machinelearning/pull/4943#discussion_r392572981) [](ancestors = 392572981)